### PR TITLE
Category and subcategory

### PR DESCRIPTION
**Pull request by avneetmalhotra**

### CODE REVIEW

File: app/models/category.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  def is_root_category?
+    root_category.nil?
+  end
+
+  def sub_category?
+    root_category.present?
+  end
+
+  def one_level_nesting
+    errors[:base] << 'Only one level nesting allowed' unless self.root_category.root_category.nil?
```
Comment (nimish13) [inline app/models/category.rb]: remove self
make use of `is_root_category?`

File: db/migrate/20171122125521_add_products_count_to_category.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddProductsCountToCategory < ActiveRecord::Migration[5.1]
+  def change
+    add_column :categories, :products_count, :integer
```
Comment (nimish13) [inline db/migrate/20171122125521_add_products_count_to_category.rb]: add default value

File: app/models/product.rb [hunk 3]
```diff
@@ -2,6 +2,10 @@ class Product < ApplicationRecord
   has_many :line_items, dependent: :restrict_with_error
   has_many :orders, through: :line_items
   has_many :carts, through: :line_items
+  has_many :categorizations
+  has_many :categories, through: :categorizations
```
Comment (nimish13) [inline app/models/product.rb]: use dependent: destroy

**Comment by avneetmalhotra** [comment]
closed in reference to #7

**Pull request closed without merging**